include/isect2d.h

```diff
@@ -2,79 +2,245 @@
+                          pairs.push_back(Pair{v[j], v[k], pairMap[key]});
+                          pairMap[key] = pairs.size()-1;
+                      }
+                  }
+              }
+          }
+          v.clear();
+      }
+   }
+
+private:
+    // from fontstash
+    static size_t hash_int(uint32_t a) {
+        a += ~(a<<15);
+        a ^=  (a>>10);
+        a +=  (a<<3);
+        a ^=  (a>>6);
+        a += ~(a<<11);
+        a ^=  (a>>16);
+        return a;
+    }
+
+    // from https://gist.github.com/badboy/6267743
+    static size_t hash_key(size_t key) {
+        key = (~key) + (key << 18);
+        key = key ^ (key >> 31);
+        key = key * 21;
+        key = key ^ (key >> 11);
+        key = key + (key << 6);
+        key = key ^ (key >> 22);
+        return key;
+    }
+
+};
+
+
+template<typename V>
+inline static std::unordered_set<std::pair<int, int>> intersect(const std::vector<AABB<V>>& _aabbs,
+                                                                V _split, V _resolution) {
+#if 0
+    std::set<std::pair<int, int>> pairs;
+
+    if (_aabbs.size() == 0) {
+        return pairs;
+    }
+
+    for (size_t i = 0; i < _aabbs.size(); ++i) {
+        for (size_t j = i + 1; j < _aabbs.size(); ++j) {
+            if (_aabbs[i].intersect(_aabbs[j])) {
+                pairs.insert({ i, j });
+            }
+        }
+    }
+
+    return pairs;
+
+#else
+
     struct AABBPair {
-        const isect2d::AABB* aabb;
+        const AABB<V>* aabb;
         unsigned int index;
     };
-    
-    std::set<std::pair<int, int>> pairs;
+
+    std::unordered_set<std::pair<int, int>> pairs;
     int n = int(_split.x * _split.y);
     std::vector<AABBPair>* gridAABBs = new std::vector<AABBPair>[n];
-    
+
     const short xpad = short(ceilf(_resolution.x / _split.x));
     const short ypad = short(ceilf(_resolution.y / _split.y));
-    
+
     short x = 0, y = 0;
 
     for (int j = 0; j < _split.y; ++j) {
         for (int i = 0; i < _split.x; ++i) {
-            isect2d::AABB cell(x, y, x + xpad, y + ypad);
+            AABB<V> cell(x, y, x + xpad, y + ypad);
 
             for (unsigned int index = 0; index < _aabbs.size(); ++index) {
-                const isect2d::AABB* aabb = &_aabbs[index];
+                const AABB<V>* aabb = &_aabbs[index];
                 // test the aabb against the current grid cell
                 if (cell.intersect(*aabb)) {
                     gridAABBs[int(i + j * _split.x)].push_back({aabb, index});
                 }
             }
             x += xpad;
-            
+
             if (x >= _resolution.x) {
                 x = 0;
                 y += ypad;
             }
         }
     }
-    
+
     for (int i = 0; i < n; ++i) {
         auto& v = gridAABBs[i];
+
         for (size_t j = 0; j < v.size(); ++j) {
             for (size_t k = j + 1; k < v.size(); ++k) {
-                if (v[j].index != v[k].index && v[j].aabb->intersect(*v[k].aabb)) {
```

Comment:
I do not remember why I added the check for index here ... it seems it was an extra redundant check.